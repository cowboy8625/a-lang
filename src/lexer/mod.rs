mod scanner;
mod span;
mod token;
mod token_stream;

use scanner::Lexer;
pub use span::Span;
pub use token::Token;
pub use token_stream::TokenStream;

pub fn lex(src: impl Into<String>) -> Result<TokenStream, Vec<String>> {
    Lexer::new(src.into().as_str()).lex().map(TokenStream::new)
}
