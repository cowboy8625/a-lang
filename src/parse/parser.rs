use super::{
    Item,
    Statement,
    LitInt,
    LitBool,
    Expr,
    Binary,
    Token,
    TokenKind
};
use std::iter::Peekable;
use std::slice::Iter;


pub struct Parser<'a> {
    stream: Peekable<Iter<'a, Token>>,
    errors: Vec<String>,
}

// declaration
// statement
// expression
// equality
// comparison
// term
// factor
// unary
// primary

impl<'a> Parser<'a> {
    pub fn new(stream: Peekable<Iter<'a, Token>>) -> Self {
        Self {
            stream,
            errors: vec![],
        }
    }

    pub fn parse(mut self) -> Result<Item, Vec<String>> {
        match self.statement() {
            Ok(item) => {
                if !self.errors.is_empty() {
                    return Err(self.errors);
                }
                Ok(item)
            }
            Err(error) => {
                self.errors.push(error);
                Err(self.errors)
            }
        }
    }

    fn match_on(&mut self, expected: &[TokenKind]) -> bool {
        let kind = self
            .stream
            .peek()
            .as_ref()
            .map(|t| t.kind())
            .unwrap_or(TokenKind::Eof);
        for token_kind in expected {
            if token_kind == &kind {
                return true;
            }
        }
        false
    }

    fn statement(&mut self) -> Result<Item, String> {
        let stmt = self.expression();
        let span = stmt.span();
        let Some(TokenKind::SemiColon) = self.stream.peek().map(|i| i.kind()) else {
            return Err(format!("Statements end in ';' {stmt:?}"));
        };
        Ok(Item::Statement(Statement {
            stmt,
            span,
        }))
    }

    fn expression(&mut self) -> Expr {
        self.comparison()
    }

    fn comparison(&mut self) -> Expr {
        use TokenKind::{EqEq, Geq, Grt, Leq, Les, Neq};
        let mut expr = self.term();
        while self.match_on(&[Grt, Les, Geq, Leq, EqEq, Neq]) {
            let op = self.stream.next().unwrap();
            let right = self.term();
            expr = Expr::from(Binary::from((expr, right, op)))
        }
        expr
    }

    fn term(&mut self) -> Expr {
        let mut expr = self.factor();
        while self.match_on(&[TokenKind::Plus, TokenKind::Minus]) {
            let op = self.stream.next().unwrap();
            let right = self.factor();
            expr = Expr::from(Binary::from((expr, right, op)))
        }
        expr
    }

    fn factor(&mut self) -> Expr {
        let mut expr = self.primary();
        while self.match_on(&[TokenKind::Star, TokenKind::Slash]) {
            let op = self.stream.next().unwrap();
            let right = self.primary();
            expr = Expr::from(Binary::from((expr, right, op)))
        }
        expr
    }

    fn primary(&mut self) -> Expr {
        self.stream
            .next()
            .and_then(|token| {
                Some(match token.kind() {
                    TokenKind::Int => Expr::from(LitInt::from(token)),
                    TokenKind::True => Expr::from(LitBool::from(token)),
                    TokenKind::False => Expr::from(LitBool::from(token)),
                    _ => unimplemented!(),
                })
            })
            .unwrap()
    }
}
