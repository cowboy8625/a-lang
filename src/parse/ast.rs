use crate::lexer::{Span, Token};
use std::str::FromStr;

#[derive(Debug, Clone)]
pub struct Type {
    pub name: Name,
    pub span: Span,
}

impl From<&Token> for Type {
    fn from(value: &Token) -> Self {
        Self {
            name: value.into(),
            span: value.span(),
        }
    }
}

#[derive(Debug, Clone)]
pub enum Item {
    Fn(ItemFn),
}

#[derive(Debug, Clone)]
pub struct Name {
    pub name: String,
    pub span: Span,
}

impl From<&Token> for Name {
    fn from(value: &Token) -> Self {
        Self {
            name: value.lexme(),
            span: value.span(),
        }
    }
}

#[derive(Debug, Clone)]
pub struct Param {
    pub name: String,
    pub kind: Type,
    pub span: Span,
}

impl From<(&Token, &Token)> for Param {
    fn from((name, kind): (&Token, &Token)) -> Self {
        let start_span = name.span();
        let end_span = kind.span();
        let span = Span::new(start_span.line, start_span.start, end_span.end);
        Self {
            name: name.lexme(),
            kind: kind.into(),
            span,
        }
    }
}

#[derive(Debug, Clone)]
pub struct ItemFn {
    pub name: Name,
    pub params: Vec<Param>,
    pub block: Block,
    pub ret_type: Option<Type>,
    pub span: Span,
}

#[derive(Debug, Clone)]
pub struct Block {
    pub stmts: Vec<Statement>,
    pub span: Span,
}

#[derive(Debug, Clone)]
pub struct Statement {
    pub stmt: Expr,
    pub span: Span,
}

#[derive(Debug, Clone)]
pub struct LitInt {
    pub value: String,
    pub span: Span,
}

impl LitInt {
    pub fn _parse<T: FromStr>(&self) -> Result<T, <T as FromStr>::Err> {
        self.value.parse::<T>()
    }
}

impl From<&Token> for LitInt {
    fn from(token: &Token) -> Self {
        Self {
            value: token.lexme().to_string(),
            span: token.span(),
        }
    }
}

#[derive(Debug, Clone)]
pub struct LitBool {
    pub value: String,
    pub span: Span,
}

impl From<&Token> for LitBool {
    fn from(token: &Token) -> Self {
        Self {
            value: token.lexme().to_string(),
            span: token.span(),
        }
    }
}

#[derive(Debug, Clone)]
pub struct Binary {
    pub left: Box<Expr>,
    pub right: Box<Expr>,
    pub op: Token,
}

impl Binary {
    pub fn span(&self) -> Span {
        let start = self.left.span();
        let end = self.right.span();
        Span::new(start.line, start.start, end.end)
    }
}

impl From<(Expr, Expr, &Token)> for Binary {
    fn from((left, right, op): (Expr, Expr, &Token)) -> Self {
        Self {
            left: Box::new(left),
            right: Box::new(right),
            op: op.clone(),
        }
    }
}

#[derive(Debug, Clone)]
pub enum Lit {
    Int(LitInt),
    Bool(LitBool),
}

impl Lit {
    pub fn span(&self) -> Span {
        match self {
            Self::Int(int) => int.span,
            Self::Bool(bl) => bl.span,
        }
    }
}

// #[derive(Debug, Clone)]
// pub struct ExprCall {
//     pub name: Name,
//     pub args: Vec<Expr>,
//     pub span: Span,
// }
//
// impl ExprCall {
//     pub fn span(&self) -> Span {
//         self.span
//     }
// }

#[derive(Debug, Clone)]
pub enum Expr {
    Lit(Lit),
    Binary(Binary),
    // Call(ExprCall),
}

impl Expr {
    pub fn span(&self) -> Span {
        match self {
            Self::Lit(lit) => lit.span(),
            Self::Binary(binary) => binary.span(),
            // Self::Call(expr_call) => expr_call.span(),
        }
    }
}

impl From<LitBool> for Expr {
    fn from(lit_bool: LitBool) -> Self {
        Self::Lit(Lit::Bool(lit_bool))
    }
}

impl From<LitInt> for Expr {
    fn from(lit_int: LitInt) -> Self {
        Self::Lit(Lit::Int(lit_int))
    }
}

impl From<Binary> for Expr {
    fn from(bin: Binary) -> Self {
        Self::Binary(bin)
    }
}
