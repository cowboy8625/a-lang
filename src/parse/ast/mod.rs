mod expr;
mod item;
pub mod keyword;
mod lit;

use crate::lexer::Span;
pub use expr::{Expr, ExprBinary, ExprCall, ExprLit, ExprVar};
pub use item::{Item, ItemFn};
pub use lit::{Lit, LitBool, LitChar, LitInt, LitStr};

#[macro_export]
macro_rules! token {
    ($name:ident) => {
        #[derive(Debug, Clone, Hash, PartialEq, Eq, PartialOrd, Ord)]
        pub struct $name {
            pub value: String,
            pub span: crate::lexer::Span,
        }

        impl $name {
            pub fn new(value: impl Into<String>, span: crate::lexer::Span) -> Self {
                Self {
                    value: value.into(),
                    span,
                }
            }
        }

        impl crate::lexer::Token for $name {
            fn new(value: String, span: crate::lexer::Span) -> Self {
                Self { value, span }
            }
            fn value(&self) -> String {
                self.value.to_string()
            }
            fn span(&self) -> crate::lexer::Span {
                self.span
            }
            fn as_any(&self) -> &dyn std::any::Any {
                self
            }
        }
    };
}

#[macro_export]
macro_rules! from_token {
    ($name:ident, $var:ident, $from:ident) => {
        impl From<$from> for $name {
            fn from(value: $from) -> Self {
                Self::$var(value)
            }
        }

        impl From<&$from> for $name {
            fn from(value: &$from) -> Self {
                Self::$var(value.clone())
            }
        }
    };
}

token!(Ident);
token!(OpAdd);
token!(OpSub);
token!(OpMul);
token!(OpDiv);
token!(OpGrt);
token!(OpLes);
token!(OpGeq);
token!(OpLeq);
token!(OpNeq);
token!(OpNot);
token!(OpEqual);
token!(OpEqualEqual);

#[derive(Debug, Clone, Hash, PartialEq, Eq, PartialOrd, Ord)]
pub enum Op {
    Add(OpAdd),
    Sub(OpSub),
    Mul(OpMul),
    Div(OpDiv),
    Grt(OpGrt),
    Les(OpLes),
    Geq(OpGeq),
    Leq(OpLeq),
    Neq(OpNeq),
    Not(OpNot),
    Equal(OpEqual),
    EqualEqual(OpEqualEqual),
}

from_token!(Op, Add, OpAdd);
from_token!(Op, Sub, OpSub);
from_token!(Op, Mul, OpMul);
from_token!(Op, Div, OpDiv);
from_token!(Op, Grt, OpGrt);
from_token!(Op, Les, OpLes);
from_token!(Op, Geq, OpGeq);
from_token!(Op, Leq, OpLeq);
from_token!(Op, Neq, OpNeq);
from_token!(Op, Not, OpNot);
from_token!(Op, Equal, OpEqual);
from_token!(Op, EqualEqual, OpEqualEqual);

token!(CtrlStar);
token!(CtrlSlash);
token!(CtrlSemiColon);
token!(CtrlColon);
token!(CtrlComma);
token!(CtrlDot);
token!(CtrlLBrace);
token!(CtrlRBrace);
token!(CtrlLBracet);
token!(CtrlRBracet);
token!(CtrlLParan);
token!(CtrlRParan);
token!(CtrlRightArrow);
token!(CtrlThickRightArrow);

#[derive(Debug, Clone, Hash, PartialEq, Eq, PartialOrd, Ord)]
pub enum Ctrl {
    Star(CtrlStar),                       // *
    Slash(CtrlSlash),                     // /
    SemiColon(CtrlSemiColon),             // ;
    Colon(CtrlColon),                     // :
    Comma(CtrlComma),                     // ,
    Dot(CtrlDot),                         // .
    LBrace(CtrlLBrace),                   // {
    RBrace(CtrlRBrace),                   // }
    LBracet(CtrlLBracet),                 // [
    RBracet(CtrlRBracet),                 // ]
    LParan(CtrlLParan),                   // (
    RParan(CtrlRParan),                   // )
    RightArrow(CtrlRightArrow),           // ->
    ThickRightArrow(CtrlThickRightArrow), // =>
}

from_token!(Ctrl, Star, CtrlStar);
from_token!(Ctrl, Slash, CtrlSlash);
from_token!(Ctrl, SemiColon, CtrlSemiColon);
from_token!(Ctrl, Colon, CtrlColon);
from_token!(Ctrl, Comma, CtrlComma);
from_token!(Ctrl, Dot, CtrlDot);
from_token!(Ctrl, LBrace, CtrlLBrace);
from_token!(Ctrl, RBrace, CtrlRBrace);
from_token!(Ctrl, LBracet, CtrlLBracet);
from_token!(Ctrl, RBracet, CtrlRBracet);
from_token!(Ctrl, LParan, CtrlLParan);
from_token!(Ctrl, RParan, CtrlRParan);
from_token!(Ctrl, RightArrow, CtrlRightArrow);
from_token!(Ctrl, ThickRightArrow, CtrlThickRightArrow);

#[derive(Debug, Clone, Hash, PartialEq, Eq, PartialOrd, Ord)]
pub struct Block {
    pub stmts: Vec<Statement>,
    pub span: Span,
}

#[derive(Debug, Clone, Hash, PartialEq, Eq, PartialOrd, Ord)]
pub struct Statement {
    pub stmt: Expr,
    pub span: Span,
}

#[derive(Debug, Clone, Hash, PartialEq, Eq, PartialOrd, Ord)]
pub struct Type(pub Ident);

impl From<&Ident> for Type {
    fn from(value: &Ident) -> Self {
        Self(value.clone())
    }
}

#[derive(Debug, Clone, Hash, PartialEq, Eq, PartialOrd, Ord)]
pub struct Param {
    pub name: Ident,
    pub kind: Type,
    pub span: Span,
}

impl From<(&Ident, &Ident)> for Param {
    fn from((name, kind): (&Ident, &Ident)) -> Self {
        let span = Span::from((name.span, kind.span));
        Self {
            name: name.clone(),
            kind: kind.into(),
            span,
        }
    }
}
