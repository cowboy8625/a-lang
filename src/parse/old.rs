use crate::lexer::{Span, Token};
use std::str::FromStr;

#[derive(Debug, Clone)]
pub struct Type {
    pub name: Name,
    pub span: Span,
}

impl From<&Token> for Type {
    fn from(value: &Token) -> Self {
        Self {
            name: value.into(),
            span: value.span(),
        }
    }
}

#[derive(Debug, Clone)]
pub enum Item {
    Fn(ItemFn),
}

#[derive(Debug, Clone)]
pub struct Name {
    pub name: String,
    pub span: Span,
}

impl From<&Token> for Name {
    fn from(value: &Token) -> Self {
        Self {
            name: value.lexme(),
            span: value.span(),
        }
    }
}

#[derive(Debug, Clone)]
pub struct Param {
    pub name: String,
    pub kind: Type,
    pub span: Span,
}

impl From<(&Token, &Token)> for Param {
    fn from((name, kind): (&Token, &Token)) -> Self {
        let start_span = name.span();
        let end_span = kind.span();
        let span = Span::new(start_span.line, start_span.start, end_span.end);
        Self {
            name: name.lexme(),
            kind: kind.into(),
            span,
        }
    }
}

#[derive(Debug, Clone)]
pub struct ItemFn {
    pub name: Name,
    pub params: Vec<Param>,
    pub block: Block,
    pub ret_type: Option<Type>,
    pub span: Span,
}

#[derive(Debug, Clone)]
pub struct Block {
    pub stmts: Vec<Statement>,
    pub span: Span,
}

#[derive(Debug, Clone)]
pub struct Statement {
    pub stmt: Expr,
    pub span: Span,
}

#[derive(Debug, Clone)]
pub struct LitInt {
    pub value: String,
    pub span: Span,
}

impl LitInt {
    pub fn _parse<T: FromStr>(&self) -> Result<T, <T as FromStr>::Err> {
        self.value.parse::<T>()
    }
}

impl From<&Token> for LitInt {
    fn from(token: &Token) -> Self {
        Self {
            value: token.lexme().to_string(),
            span: token.span(),
        }
    }
}

#[derive(Debug, Clone)]
pub struct LitBool {
    pub value: String,
    pub span: Span,
}

impl From<&Token> for LitBool {
    fn from(token: &Token) -> Self {
        Self {
            value: token.lexme().to_string(),
            span: token.span(),
        }
    }
}

#[derive(Debug, Clone)]
pub struct Binary {
    pub left: Box<Expr>,
    pub right: Box<Expr>,
    pub op: Token,
}

impl Binary {
    pub fn span(&self) -> Span {
        let start = self.left.span();
        let end = self.right.span();
        Span::new(start.line, start.start, end.end)
    }
}

impl From<(Expr, Expr, &Token)> for Binary {
    fn from((left, right, op): (Expr, Expr, &Token)) -> Self {
        Self {
            left: Box::new(left),
            right: Box::new(right),
            op: op.clone(),
        }
    }
}

#[derive(Debug, Clone)]
pub enum Lit {
    Int(LitInt),
    Bool(LitBool),
}

impl Lit {
    pub fn span(&self) -> Span {
        match self {
            Self::Int(int) => int.span,
            Self::Bool(bl) => bl.span,
        }
    }
}

#[derive(Debug, Clone)]
pub struct ExprCall {
    pub caller: Box<Expr>,
    pub args: Vec<Expr>,
    pub span: Span,
}

impl ExprCall {
    pub fn span(&self) -> Span {
        self.span
    }
}

#[derive(Debug, Clone)]
pub struct Var {
    pub name: Name,
    pub span: Span,
}

impl Var {
    pub fn span(&self) -> Span {
        self.span
    }
}

impl From<&Token> for Var {
    fn from(value: &Token) -> Self {
        Self {
            name: value.into(),
            span: value.span(),
        }
    }
}

#[derive(Debug, Clone)]
pub enum Expr {
    Lit(Lit),
    Binary(Binary),
    Call(ExprCall),
    Var(Var),
}

impl Expr {
    pub fn span(&self) -> Span {
        match self {
            Self::Lit(i) => i.span(),
            Self::Binary(i) => i.span(),
            Self::Call(i) => i.span(),
            Self::Var(i) => i.span(),
        }
    }
}

impl From<LitBool> for Expr {
    fn from(lit_bool: LitBool) -> Self {
        Self::Lit(Lit::Bool(lit_bool))
    }
}

impl From<LitInt> for Expr {
    fn from(lit_int: LitInt) -> Self {
        Self::Lit(Lit::Int(lit_int))
    }
}

impl From<Binary> for Expr {
    fn from(bin: Binary) -> Self {
        Self::Binary(bin)
    }
}

impl From<Var> for Expr {
    fn from(var: Var) -> Self {
        Self::Var(var)
    }
}
