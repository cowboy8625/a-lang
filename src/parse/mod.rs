#![allow(unused)]
mod ast;
// mod new_parser;
mod parser;

pub use ast::keyword;
pub use ast::*;

use parser::Parser;

use crate::lexer::{Token, TokenStream};

pub fn parse(stream: TokenStream) -> Result<Vec<Item>, Vec<String>> {
    // new_parser::parse(stream)
    Parser::new(stream).parse()
}
