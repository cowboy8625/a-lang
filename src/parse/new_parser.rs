use super::ast::*;
use crate::lexer::TokenStream;


// declaration
// statement
// expression
// equality
// comparison
// term
// factor
// unary
// primary

type Result<T> = std::result::Result<T, String>;

struct ParseStream {
    ast: Vec<Box<dyn Parse>>
}

impl Parser for ParseStream {
    fn parse(self, tokens: TokenStream) -> Result<Self::Output> {
        
    }
}

trait Parser: Sized {
    type Output;
    fn parse(self, tokens: TokenStream) -> Result<Self::Output>;
}

trait Parse: Sized {
    fn parse(tokens: &mut TokenStream) -> Result<Self>;
}



impl Parse for Lit {
    fn parse(tokens: &mut TokenStream) -> Result<Self> {
        if tokens.peek::<LitInt>().is_some() {
            return tokens
                .next_as::<LitInt>()
                .map(Lit::from)
                .ok_or::<String>("not a lit".into());
        } else if tokens.peek::<LitBool>().is_some() {
            return tokens
                .next_as::<LitBool>()
                .map(Lit::from)
                .ok_or::<String>("not a lit".into());
        } else if tokens.peek::<LitStr>().is_some() {
            return tokens
                .next_as::<LitStr>()
                .map(Lit::from)
                .ok_or::<String>("not a lit".into());
        } else if tokens.peek::<LitChar>().is_some() {
            return tokens
                .next_as::<LitStr>()
                .map(Lit::from)
                .ok_or::<String>("not a lit".into());
        }
        Err("not a lit".into())
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::lexer::{lex, Span};
    use pretty_assertions::assert_eq;
    #[test]
    fn test_lit() {
        let mut tokens = lex("1").unwrap();
        let left = Lit::parse(&mut tokens).unwrap();
        let right = Lit::Int(LitInt::new("1", Span::new(0,0,1)));
        assert_eq!(left, right);
    }
}


pub fn parse(tokens: TokenStream) -> std::result::Result<Vec<Item>, Vec<String>> {
    todo!()
}

// type ParseResult<T> = Result<(T, TokenStream), (TokenStream, String)>;

// fn parse_factor(mut tokens: TokenStream) -> ParseResult<ExprBinary> {
//     if tokens.peek_nth::<OpMul>(2).is_none() || tokens.peek_nth::<OpDiv>(2).is_none() {
//         return Err((tokens, "Not a factor bin".into()));
//     }
//     let (lhs, tokens) = parse_expr_lit(tokens)?;
//     let op = Op::from(tokens.next());
//     let (rhs, tokens) = parse_expr_lit(tokens)?;
//     // OpMul, OpDiv
//     todo!()
// }
//
// fn parse_expr_lit(mut tokens: TokenStream) -> ParseResult<ExprLit> {
//     parse_lit(tokens)
//         .map(|(i, t)| (ExprLit::from(i), t))
// }
//
// fn parse_lit(mut tokens: TokenStream) -> ParseResult<Lit> {
//     parse_lit_int(tokens)
//         .map(|(i, t)| (Lit::from(i), t))
//         .or_else(|(t, _)| parse_lit_bool(t)
//                  .map(|(i, t)| (Lit::from(i), t)))
//         .or_else(|(t, _)| parse_lit_str(t)
//                  .map(|(i, t)| (Lit::from(i), t)))
//         .or_else(|(t, _)| parse_lit_char(t)
//                  .map(|(i, t)| (Lit::from(i), t)))
// }
//
// fn parse_lit_int(mut tokens: TokenStream) -> ParseResult<LitInt> {
//     let peek = tokens.peek::<LitInt>().cloned();
//     let Some(litint) = peek else {
//         return Err((tokens, "not a litint".into()));
//     };
//     tokens.next();
//     Ok((litint.clone(), tokens))
// }
//
// fn parse_lit_bool(mut tokens: TokenStream) -> ParseResult<LitBool> {
//     let peek = tokens.peek::<LitBool>().cloned();
//     let Some(litbool) = peek else {
//         return Err((tokens, "not a lit bool".into()));
//     };
//     tokens.next();
//     Ok((litbool.clone(), tokens))
// }
//
// fn parse_lit_str(mut tokens: TokenStream) -> ParseResult<LitStr> {
//     let peek = tokens.peek::<LitStr>().cloned();
//     let Some(litstr) = peek else {
//         return Err((tokens, "not a lit str".into()));
//     };
//     tokens.next();
//     Ok((litstr.clone(), tokens))
// }
//
// fn parse_lit_char(mut tokens: TokenStream) -> ParseResult<LitChar> {
//     let peek = tokens.peek::<LitChar>().cloned();
//     let Some(litchar) = peek else {
//         return Err((tokens, "not a litchar".into()));
//     };
//     tokens.next();
//     Ok((litchar.clone(), tokens))
// }
//
// #[cfg(test)]
// mod tests {
//     use super::*;
//     use crate::lexer::{lex, Span};
//     use pretty_assertions::assert_eq;
//
//     fn setup(input: &str) -> String {
//         lex(input)
//             .and_then(parse)
//             .unwrap()
//             .iter()
//             .map(ToString::to_string)
//             .collect::<String>()
//     }
//     #[test]
//     fn parse_binary() {
//         let src = "1 + 2 * 3";
//         let left = setup(src);
//         let right = "(+ 1 (* 2 3))";
//         assert_eq!(left, right);
//     }
//
//     #[test]
//     fn test_lit_int() {
//         let src = "1";
//         let tokens = lex(src).unwrap();
//         let (left, _) = parse_lit_int(tokens).unwrap();
//         let right = LitInt::new("1", Span::new(0, 0, 1));
//         assert_eq!(left, right);
//     }
//
//     #[test]
//     fn test_lit_bool() {
//         let src = "true";
//         let tokens = lex(src).unwrap();
//         let (left, _) = parse_lit_bool(tokens).unwrap();
//         let right = LitBool::new("true", Span::new(0, 0, 4));
//         assert_eq!(left, right);
//     }
//
//     #[test]
//     fn test_lit_str() {
//         let src = r#""Hello World""#;
//         let tokens = lex(src).unwrap();
//         let (left, _) = parse_lit_str(tokens).unwrap();
//         let right = LitStr::new("Hello World", Span::new(0, 0, 13));
//         assert_eq!(left, right);
//     }
//
//     #[test]
//     fn test_lit_char() {
//         let src = "'H'";
//         let tokens = lex(src).unwrap();
//         let (left, _) = parse_lit_char(tokens).unwrap();
//         let right = LitChar::new("H", Span::new(0, 0, 3));
//         assert_eq!(left, right);
//     }
//
//     #[test]
//     fn test_lit() {
//         let src = "1";
//         let tokens = lex(src).unwrap();
//         let (left, _) = parse_lit(tokens).unwrap();
//         let right = Lit::from(LitInt::new("1", Span::new(0, 0, 1)));
//         assert_eq!(left, right, "Lit LitInt");
//
//         let src = "true";
//         let tokens = lex(src).unwrap();
//         let (left, _) = parse_lit(tokens).unwrap();
//         let right = Lit::from(LitBool::new("true", Span::new(0, 0, 4)));
//         assert_eq!(left, right, "Lit LitBool");
//
//         let src = r#""Hello World""#;
//         let tokens = lex(src).unwrap();
//         let (left, _) = parse_lit(tokens).unwrap();
//         let right = Lit::from(LitStr::new("Hello World", Span::new(0, 0, 13)));
//         assert_eq!(left, right, "Lit LitStr");
//
//         let src = "'H'";
//         let tokens = lex(src).unwrap();
//         let (left, _) = parse_lit(tokens).unwrap();
//         let right = Lit::from(LitChar::new("H", Span::new(0, 0, 3)));
//         assert_eq!(left, right, "Lit LitChar");
//     }
//
//     #[test]
//     fn test_expr_lit() {
//         let src = "1";
//         let tokens = lex(src).unwrap();
//         let (left, _) = parse_expr_lit(tokens).unwrap();
//         let right = ExprLit::from(Lit::from(LitInt::new("1", Span::new(0, 0, 1))));
//         assert_eq!(left, right, "ExprLit Lit LitInt");
//
//         let src = "true";
//         let tokens = lex(src).unwrap();
//         let (left, _) = parse_expr_lit(tokens).unwrap();
//         let right = ExprLit::from(Lit::from(LitBool::new("true", Span::new(0, 0, 4))));
//         assert_eq!(left, right, "ExprLit Lit LitBool");
//
//         let src = r#""Hello World""#;
//         let tokens = lex(src).unwrap();
//         let (left, _) = parse_expr_lit(tokens).unwrap();
//         let right = ExprLit::from(Lit::from(LitStr::new("Hello World", Span::new(0, 0, 13))));
//         assert_eq!(left, right, "ExprLit Lit LitStr");
//
//         let src = "'H'";
//         let tokens = lex(src).unwrap();
//         let (left, _) = parse_expr_lit(tokens).unwrap();
//         let right = ExprLit::from(Lit::from(LitChar::new("H", Span::new(0, 0, 3))));
//         assert_eq!(left, right, "ExprLit Lit LitChar");
//     }
// }
