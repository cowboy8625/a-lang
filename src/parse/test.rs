use super::*;
use crate::lexer::lex;
use pretty_assertions::assert_eq;

macro_rules! snapshot {
    ($name:tt, $path:tt) => {
        #[test]
        fn $name() {
            use super::parse;
            use crate::lexer::lex;
            let contents = include_str!($path);
            let tokens = lex(contents).unwrap();
            let ast = parse(tokens).unwrap();
            let ast_string = ast
                .iter()
                .map(|node| format!("{node}\n"))
                .collect::<String>();
            let mut settings = insta::Settings::clone_current();
            settings.set_snapshot_path("testdata/output/");
            settings.bind(|| {
                insta::assert_snapshot!(ast_string);
            });
        }
    };
}

snapshot!(parse_binary, "testdata/snapshots/binary.a");
snapshot!(parse_ifelse, "testdata/snapshots/ifelse.a");
