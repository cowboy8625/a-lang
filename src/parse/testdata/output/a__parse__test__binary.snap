---
source: src/parse/test.rs
expression: ast_string
---
(func main <NULL> () (return (+ 1 (* 2 3)))
)

